keps/sig-apps/4212-improve-node-maintenance/README.md

```diff
@@ -0,0 +1,1065 @@
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the TargetedByNodeMaintenance
+  condition.
+- Introduce a node maintenance period or a TTL optional field as an upper bound on the duration of
+  node maintenance. Then the node maintenance would be garbage collected and the node made
+  schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from a given node. This should complement the node's `.spec.unschedulable`,
+which prevents new pods from being scheduled on such a node. We will focus primarily on
+`kubectl drain`, but the API can also be used by other drain implementations as well (e.g. node
+autoscalers).
+
+We will first introduce the NodeMaintenance API and then later modify the behavior of the
+Kubernetes system to fix all the node drain issues we mentioned earlier. The name, NodeMaintenance
+is subject to change; please see the alternatives section for other suggestions.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration with a condition. The pods would be selected according to
+the node at first, but the selection mechanism can be extended later. Controllers can then implement
+the migration. The advantage of this approach is that controllers do not have to be aware of the
+NodeMaintenance object (no RBAC changes required). They only have to observe pods they own and react
+by migrating them. The first candidate is a deployment controller, since its workloads support
+surging to another nodes. This would help to eliminate downtime not only for single replica
+applications, but for HA applications as well. 
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I also want to observe the node drain via the API and check on
+its progress. I also want to be able to discover workloads that are blocking the node drain.
```

Comment:
We would only report the summary in `.status.targetedPods`. For the rest, the user would have to filter out all the pods that have `TargetedByNodeMaintenance` condition. (We really should come up with a better name for the condition).

Other tracking or required information could be added to the NodeMainteannce status later if needed.

Reply:
> While it could be out of scope of this kep: With this change I still not see a good way of determening the drain status.
How does an admin know if all relevant pods were evicted - or not?

Currently in the KEP, we are marking all the pods, knowing that not all of them will be deleted (see drain filters). This is problematic. We should discuss the selection and reporting logic with wider audience.  

- Should we target all pods or just the ones we know can be deleted (drain filters)? The later will allow us to incrementally add features into NM, such as support for DaemonSets. And be compatible with `kubectl drain`. The first approach would allow third-party applications to react to the pods we are unsure about.
- If we target all the pods, should we report the expected pods to be deleted in another status field?

Reply:
You / we can mark than in an UNRESOLVED section and save that for a decision after the first alpha code merges.

Reply:
Ack, I have added these question under the UNRESOLVED block to the `NodeMaintenance Controller` section. And I have added another todo under the `NodeMaintenance API` section as well.